### PR TITLE
rqt_plot: Save and restore axes settings. Fixes #234

### LINKED ISSUE #234: rqt_plot: save and restore axes scaling
putting this on my todo list; it would be nice if rqt_plot or the underlying plotting widget could save and restore the axes scale or limits, so that the plot effectively remembers its size between runs.

### PR DESCRIPTION
**Pull request by trainman419**

### CODE REVIEW

**Comment by trainman419** [comment]
I'm able to reproduce this on my machine by opening and closing rqt_plot with PyQtGraph, but not with MatPlot. It looks like this is due to a difference in the way that each backend returns the plot limits; PyQtGraph returns the plot limits as a normal list, and they're stored in the settings file like this:

```
Default\pluginmanager\plugin__rqt_plot__Plot__1\plugin\y_limits=0, 1
```

When this is restored, it's reinterpreted as a list of strings instead of as a list of integers.

Meanwhile, MatPlot returns the limits as a numpy array, and they're stored differently:

```
Default\pluginmanager\plugin__rqt_plot__Plot__1\plugin\y_limits=@Variant(\0\0\0\x7f\0\0\0\xePyQt_PyObject\0\0\0\0\x99\x63numpy.core.multiarray\nscalar\np0\n(cnumpy\ndtype\np1\n(S'f8'\np2\nI0\nI1\ntp3\nRp4\n(I3\nS'<'\np5\nNNNI-1\nI-1\nI0\ntp6\nbS'\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00'\np7\ntp8\nRp9\n.), @Variant(\0\0\0\x7f\0\0\0\xePyQt_PyObject\0\0\0\0\x96\x63numpy.core.multiarray\nscalar\np0\n(cnumpy\ndtype\np1\n(S'f8'\np2\nI0\nI1\ntp3\nRp4\n(I3\nS'<'\np5\nNNNI-1\nI-1\nI0\ntp6\nbS'\\x00\\x00\\x00\\x00\\x00\\x00\\xf0?'\np7\ntp8\nRp9\n.)
```

Since this appears to be explicitly encoding the type, it gets loaded properly when restored.

I'll work on standardizing the output type, and I'll add a bit more code to explicitly convert the loaded values to floats before setting the limits; that should fix this.

**Comment by trainman419** [comment]
Fixed the formatting for saving and loading limits in a1f9dd8 ; please test again.

**Comment by vincentrou** [comment]
It works fine for me now !

**Comment by trainman419** [comment]
@ablasdel or @dirk-thomas do you have time to review and merge this?

File: rqt_plot/src/rqt_plot/data_plot/__init__.py [hunk 1]
```diff
@@ -210,12 +211,37 @@ def save_settings(self, plugin_settings, instance_settings):
         Currently, this is just the plot type, but may include more useful
         data in the future"""
         instance_settings.set_value('plot_type', self._plot_index)
+        xlim = self.get_xlim()
+        ylim = self.get_ylim()
+        # convert limits to normal arrays of floats; some backends return numpy
+        # arrays
+        xlim = [ float(x) for x in xlim ]
+        ylim = [ float(y) for y in ylim ]
```
Comment (dirk-thomas) [inline rqt_plot/src/rqt_plot/data_plot/__init__.py]: PEP8: no spaces inside brackets.

File: rqt_plot/src/rqt_plot/data_plot/__init__.py [hunk 2]
```diff
@@ -210,12 +211,37 @@ def save_settings(self, plugin_settings, instance_settings):
...
         self._switch_data_plot_widget(int(instance_settings.value('plot_type', 0)))
+        xlim = unpack(instance_settings.value('x_limits', []))
+        ylim = unpack(instance_settings.value('y_limits', []))
+        if xlim:
+            # convert limits to an array of floats; they're often lists of
+            # strings
+            try:
+                xlim = [ float(x) for x in xlim ]
+                self.set_xlim(xlim)
+            except:
```
Comment (dirk-thomas) [inline rqt_plot/src/rqt_plot/data_plot/__init__.py]: Is there no specific exception which can be catched here?

Comment (trainman419) [inline rqt_plot/src/rqt_plot/data_plot/__init__.py]: The set_xlim call might throw a number of different exceptions depending on which backend is in use.

I'm now aware of any exception which shouldn't be caught here, and the fall-back behavior is reasonable in all cases.

**Comment by trainman419** [comment]
I've fixed the pep8 formatting issues.

**Comment by trainman419** [comment]
Bump?

**Comment by ablasdel** [comment]
I will take some time this afternoon to test this out.

**Comment by ablasdel** [comment]
Hey would you mind maybe rebasing to the current master(groovy-devel) head and maybe squash this up? 
It won't apply to my fresh checkout for some reason and I can't figure out why.

```
error: patch failed: rqt_plot/src/rqt_plot/data_plot/__init__.py:36
error: rqt_plot/src/rqt_plot/data_plot/__init__.py: patch does not apply
error: patch failed: rqt_plot/src/rqt_plot/data_plot/__init__.py:221
error: rqt_plot/src/rqt_plot/data_plot/__init__.py: patch does not apply
error: patch failed: rqt_plot/src/rqt_plot/data_plot/__init__.py:211
error: rqt_plot/src/rqt_plot/data_plot/__init__.py: patch does not apply
error: patch failed: rqt_plot/src/rqt_plot/data_plot/__init__.py:215
error: rqt_plot/src/rqt_plot/data_plot/__init__.py: patch does not apply
```

**Comment by trainman419** [comment]
Github indicated that it would apply cleanly, and I was able to apply it cleanly, but I've rebased and squashed anyway.

**Comment by ablasdel** [comment]
It is possible I am doing something crazy with my patch apply steps but this new rebased/squashed patch applies cleanly with the same process so I'm testing now. Thanks!

**Comment by ablasdel** [comment]
I tested with the normal data I get from my sensors here and some topics publishing a sin wave. It seems to save/restore the bounds correctly and cleanly. @dirk-thomas do you want to test this out too before merging?

**Comment by dirk-thomas** [comment]
I don't have much sparse cycles to give it a try - I trust you guys to go ahead with the merge.

I would follow the motto: "Any contribution is a good one compared to having a stall project"

**Comment by ablasdel** [comment]
Agreed. I'm merging this now.

**Pull request merged**